### DISCUSSION TITLE
Help me install NextLS in VS Code

### DISCUSSION THREAD

**Discussion by lud-wj** [elixir-tools/discussions#5] [Q&A] [answered]
Hello, 

I'm trying to make it work. I disabled ElixirLS and installed the extension from the VS Code panel.

This is the output I get in VS Code :

```
[Error - 15:02:24] Stopping server failed
Error: Stopping the server timed out
    at /home/lud/.vscode/extensions/elixir-tools.elixir-tools-0.2.2/dist/extension.js:2:150423
```

(duplicated several times).

I have those settings in VSC:

```
  "elixir-tools.nextls.enable": true,
  "elixir-tools.credo.enable": false,
```

I am runnin Xubuntu with :

```
Erlang/OTP 25 [erts-13.0] [source] [64-bit] [smp:12:12] [ds:12:12:10] [async-threads:1] [jit:ns]

Elixir 1.14.0 (compiled with Erlang/OTP 25)
```

Thank you :)

**Comment by mhanberg** [comment]
Can you try (if you already haven't) quitting vscode and re-opening?

I have seen some weirdness with the first time initialization on VSCode and I'm not sure why yet. 

And did you enable it through the settings GUI or in a file?

**Reply by lud-wj** [reply]
Hi!

I retried it to see, with even a `killall code` which, surprisingly, stopped something …, but I have the same error.

I also have a notification: "connection to server is erroring, Shutting down server.".

I enabled just by adding   `"elixir-tools.nextls.enable": true,` in the settings JSON file. I'll try with the UI if I can find where that is.

**Reply by lud-wj** [reply]
I'm trying with the UI but it does not adds more output when I turn it on. Reloading the vscode window triggers the errors again.

**Reply by mhanberg** [reply]
Does this return valid output? and how long did it take?

`curl --silent -L -H "Accept: application/vnd.github+json" -H "X-GitHub-Api-Version: 2022-11-28" https://api.github.com/repos/elixir-tools/next-ls/releases/latest`

also, what happens if you run this `NEXTLS_VERSION=0.5.1 /home/lud/.vscode/extensions/elixir-tools.elixir-tools-0.2.2/bin/nextls --stdio` in your project directory? (it should just sit there)

**Reply by lud-wj** [reply]
The first thing returns a huge JSON instantly.

The manual start of the server changes nothing infortunately. It seems to run properly because the command does not return, and I can Ctrl+C twice to exit the beam runtime. But no changes on VS Code.

**Reply by lud-wj** [reply]
If I start without `--stdio` but with `--port 9999` and set the adapter to "tcp" in vscode with the same port, I get this output in VS Code:

```
[NextLS] NextLS v0.5.1 has initialized!
[NextLS] Booting runime...
[NextLS] ** (CompileError) config/runtime.exs:2: module Dotenvy is not loaded and could not be found
    (elixir 1.14.0) lib/code.ex:422: Code.validated_eval_string/3
```

(not sure if you use Dotenvy, but my project does, so a problem with deps perhaps)

**Reply by mhanberg** [reply]
> But no changes on VS Code.

I just wanted to make sure that the program was able to start at all.

Now can you run the `NEXTLS_VERSION=0.5.1 /home/lud/.vscode/extensions/elixir-tools.elixir-tools-0.2.2/bin/nextls --port 9000` and in the vscode settings, set nextls to use TCP and port 9000, then quit and reopen (also, make sure neither program is running, then start the nextls command, then open vscode)

**Reply by mhanberg** [reply]
Oh, i missed your last message, you're ahead of the game!

Let me look into that.

**Reply by lud-wj** [reply]
Yep I just did (with 9999), please see above. I'll retry with a clean start

**Reply by lud-wj** [reply]
I have some errors in the server

```
15:47:16.452 [error] GenServer #PID<0.219.0> terminating
** (FunctionClauseError) no function clause matching in NextLS.Runtime.handle_info/2
    (next_ls 0.5.1) lib/next_ls/runtime.ex:156: NextLS.Runtime.handle_info({:DOWN, #Reference<0.2785840508.247726084.30186>, :port, #Port<0.10>, :normal}, %{compiler_ref: nil, errors: nil, extension_registry: NextLS.ExtensionRegistry, logger: #PID<0.214.0>, parent: #PID<0.213.0>, port: #Port<0.10>, task_supervisor: :runtime_task_supervisor})
    (stdlib 4.0) gen_server.erl:1120: :gen_server.try_dispatch/4
    (stdlib 4.0) gen_server.erl:1197: :gen_server.handle_msg/6
    (stdlib 4.0) proc_lib.erl:240: :proc_lib.init_p_do_apply/3
Last message: {:DOWN, #Reference<0.2785840508.247726084.30186>, :port, #Port<0.10>, :normal}

15:47:16.476 [error] Task #PID<0.221.0> started from #PID<0.213.0> terminating
** (stop) exited in: GenServer.call(#PID<0.219.0>, :ready?, 5000)
    ** (EXIT) an exception was raised:
        ** (FunctionClauseError) no function clause matching in NextLS.Runtime.handle_info/2
            (next_ls 0.5.1) lib/next_ls/runtime.ex:156: NextLS.Runtime.handle_info({:DOWN, #Reference<0.2785840508.247726084.30186>, :port, #Port<0.10>, :normal}, %{compiler_ref: nil, errors: nil, extension_registry: NextLS.ExtensionRegistry, logger: #PID<0.214.0>, parent: #PID<0.213.0>, port: #Port<0.10>, task_supervisor: :runtime_task_supervisor})
            (stdlib 4.0) gen_server.erl:1120: :gen_server.try_dispatch/4
            (stdlib 4.0) gen_server.erl:1197: :gen_server.handle_msg/6
            (stdlib 4.0) proc_lib.erl:240: :proc_lib.init_p_do_apply/3
    (elixir 1.14.0) lib/gen_server.ex:1038: GenServer.call/3
    (next_ls 0.5.1) lib/next_ls.ex:474: NextLS.wait_until/2
    (next_ls 0.5.1) lib/next_ls.ex:292: anonymous fn/2 in NextLS.handle_notification/2
    (elixir 1.14.0) lib/task/supervised.ex:89: Task.Supervised.invoke_mfa/2
    (elixir 1.14.0) lib/task/supervised.ex:34: Task.Supervised.reply/4
    (stdlib 4.0) proc_lib.erl:240: :proc_lib.init_p_do_apply/3
Function: #Function<10.99235021/0 in NextLS.handle_notification/2>
    Args: []

```

Maybe it does not handle to use deps in the runtime.exs file ?

**Reply by mhanberg** [reply]
I suspect this has to do with how next-ls starts the runtime, and your config/runtime.exs file.

**Reply by mhanberg** [reply]
man, I am really missing your messages here.

I would have expected the VSCode extension to have printed that error, but i'll try and reproduce

My work project uses releases.exs, not runtime.exs, so that is probably why i never noticed it

**Reply by lud-wj** [reply]
I answered in a new thread :)

**Comment by lud-wj** [comment]
I cleaned up my `runtime.exs` to use a `fn` instead of calling the dependency, but it seems to have the same problem.

IIRC `runtime.exs` is called in all situations whereas `releases.exs` is never called during development. It seems that NextLS tries to execute my config file:
(using `System.fetch_env!` instead of `Dotenvy.env!`)
```
[NextLS] NextLS v0.5.1 has initialized!
[NextLS] Booting runime...
[NextLS] ** (System.EnvError) could not fetch environment variable "DB_USERNAME" because it is not set
    (elixir 1.14.0) lib/system.ex:706: System.fetch_env!/1
```

**Reply by lud-wj** [reply]
Ok so if I wrap all the `runtime.exs` content in `if false do` then the NextLS output is displaying the compilation log, so it seems that it boots fine.

I reverted to `stdio` and that works too.

**Comment by mhanberg** [comment, marked as answer]
I think we've determined that your particular issue is, in my opinion, a bug in NextLS, I have created an issue there: https://github.com/elixir-tools/next-ls/issues/97

Sorry about that! I'll let you know when it is fixed and released.

**Reply by lud-wj** [reply]
Nice ! Thank you :)
